src/combinator.rs

```diff
@@ -260,30 +261,28 @@ where
         let span = inp.span_since(&before);
         let new_alt = inp.errors.alt.take();
 
+        inp.errors.alt = old_alt;
         match res {
             Ok(out) => {
                 if (self.filter)(&out) {
                     // If successful, reinsert the original alt and then apply the new alt on top of it, since both are valid
-                    inp.errors.alt = old_alt;
                     if let Some(new_alt) = new_alt {
                         inp.add_alt_err(&new_alt.pos, new_alt.err);
                     }
                     Ok(M::bind(|| out))
                 } else {
                     // If unsuccessful, reinsert the original alt but replace the new alt with the "something else" error (since it overrides it)
                     let expected = [DefaultExpected::SomethingElse];
-                    let err = E::Error::expected_found(expected, None, span);
-                    inp.errors.alt = old_alt;
+                    let err = E::Error::expected_found(expected, found, span);
```

Comment:
Hmm. I do understand the reasoning here: and I think it's the best we can do for now. But it isn't ideal long-term. Ideally we'd point at the *whole* of the matched input here instead of just the first token. Do you think you could add a TODO noting that this is something to improve on later?